### PR TITLE
Blocking functions

### PR DESCRIPTION
**Pull request by nbgl**
Defines what a blocking function is, and implements some blocking functions and higher order blocking functions.

Implemented are three blocking functions:

1. LSH blocking (blocking on random bits),
2. Blocking on continuous variables,
3. Blocking on precomputed blocks (this is just for convenience).

The higher order blocking functions are conjunction and disjunction.

### CODE REVIEW

**Comment by nbgl** [comment]
It has come to my attention that this fails on Python 3.4 because of type annotations. Hence why Jenkins is failing.

My proposal is to end support for Python 3.4… 😃

File: anonlink/blocking.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+    """
+    __slots__ = '_funcs'
+
+    @overload
+    def __init__(self,
+                 funcs: Iterable[BlockingFunction]
+                 ) -> None:
+        pass
+
+    @overload
```
Comment (hardbyte) [inline anonlink/blocking.py]: Well that is cool, I did not know this thing. Just to help future contributors can you add a small comment with link.

File: anonlink/blocking.py [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+                 ) -> None:
+        pass
+
+    def __init__(self, *args):
+        if len(args) == 0:
+            raise ValueError('at least one argument required')
+        elif len(args) == 1:
+            self._funcs = tuple(args[0])
+        else:
+            self._funcs = args
```
Comment (hardbyte) [inline anonlink/blocking.py]: The fancy cool overlead type signatures and `__init__` method are entirely the same in these - so this should be shared code.

File: anonlink/blocking.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+from random import Random
+from typing import (Any, Callable, Hashable, Iterable, List,
+                    Optional, overload, Sequence, Tuple, TypeVar, Union)
+
+
+BlockingFunction = Callable[[int, int, Sequence[bool]], Iterable[Hashable]]
+
+
+_T = TypeVar('_T')
+def _evalf(__funcs : Iterable[Callable[..., _T]],
```
Comment (hardbyte) [inline anonlink/blocking.py]: Docstring or comment required here

File: anonlink/blocking.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+
+class bit_blocking:
+    """ Blocking on bits of the hash.
+
+        This blocking type has the advantage that it doesn't require any
+        auxilliary information. Despite this, it is still probabilistic
+        and may decrease recall.
+
+        :param int r: The number of bits to use per block.
+        :param int g: The number of blocks per record.
```
Comment (hardbyte) [inline anonlink/blocking.py]: Add docstring for `seed`

**Comment by hardbyte** [comment]
Agree with dropping support for Python versions of old. Let's just support Python 3.6.

File: anonlink/blocking.py (hunk 2 above)
Comment (nbgl) [reply anonlink/blocking.py]: Done with an abstract base class.

File: anonlink/blocking.py (hunk 1 above)
Comment (nbgl) [reply anonlink/blocking.py]: Done! Link to http://mypy.readthedocs.io/en/latest/function_overloading.html

File: anonlink/blocking.py (hunk 3 above)
Comment (nbgl) [reply anonlink/blocking.py]: Sure!

File: anonlink/blocking.py (hunk 4 above)
Comment (nbgl) [reply anonlink/blocking.py]: Thanks for this. I forgot about it

File: anonlink/blocking.py [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+BlockingFunction = Callable[[int, int, Sequence[bool]], Iterable[Hashable]]
+
+
+_T = TypeVar('_T')
+def _evalf(__funcs : 'Iterable[Callable[..., _T]]',  # https://github.com/python/typing/issues/259
+           *args,
+           **kwargs) -> Iterable[_T]:
+    """ Apply the same arguments to a number of functions.
+
+        This of this as reverse map. Instead of having one function and
```
Comment (hardbyte) [inline anonlink/blocking.py]: Typo here. "This of this"

File: tests/test_blocking.py [hunk 6]
```diff
@@ -0,0 +1,225 @@
+import itertools
+import operator
+import random
+import unittest
+
+from anonlink import blocking
+
+
```
Comment (unzvfu) [inline tests/test_blocking.py]: Is it worth parameterising these tests by `datasets` and `samples` to allow easier checking of edge cases?

File: anonlink/blocking.py [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+
+
+BlockingFunction = Callable[[int, int, Sequence[bool]], Iterable[Hashable]]
+
+
+_T = TypeVar('_T')
+def _evalf(__funcs : 'Iterable[Callable[..., _T]]',  # https://github.com/python/typing/issues/259
+           *args,
+           **kwargs) -> Iterable[_T]:
+    """ Apply the same arguments to a number of functions.
```
Comment (unzvfu) [inline anonlink/blocking.py]: "Apply a number of functions to the same arguments."

File: tests/test_blocking.py (hunk 6 above)
Comment (nbgl) [reply tests/test_blocking.py]: Definitely! I don’t want to do it now (I’d like to start integrating with the entity service on Monday and I have a bunch more work to do before then) but I’ll do it before merging api-new to master.

File: anonlink/blocking.py (hunk 7 above)
Comment (nbgl) [reply anonlink/blocking.py]: Thanks, that does make more sense

File: anonlink/blocking.py [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+            msg = 'r is expected to be positive but is {}'.format(r)
+            raise ValueError(msg)
+        
+        self._initial_params = g, r, seed  # type: Optional[Tuple[int, int, Any]]
+        self._hash_len = None  # type: Optional[int]
+        self._hash_indices = None  # type: Optional[Sequence[Sequence[int]]]
+
+    def _first_call(self,
+                    hash_len: int
+                    ) -> None:
```
Comment (unzvfu) [inline anonlink/blocking.py]: I don't understand the point of this function. Why not just calculate `_hash_indices` in the constructor? That way you don't need the `_initial_params` slot at all, and you can delete this function.

Oh, I see, it's because you set `_hash_len` based on the length of the first hash you receive.  I don't much like that style. The hash length will be a system parameter; better (and much simpler) to pass it as a parameter to the constructor.

Comment (nbgl) [reply anonlink/blocking.py]: I agree that the implementation would be simpler if we took the hash length as a parameter. However, I’d like to avoid the user of the API having to specify the hash length. This would lead to more complexity in their code.

It makes the API easier to use when it just infers whatever it needs to.

Comment (unzvfu) [reply anonlink/blocking.py]: Your call.

Personally I think the additional convenience to the caller is pretty minor in this case, while the inconvenience for the maintainer is, let's say, non-trivial.

I'm also not convinced that hiding the dependency on the caller's hash length is an advantage for the caller. For example, I think the caller is more likely to be surprised by the ValueError on L167 if they haven't had to acknowledge the dependence on the hash length.

**Comment by nbgl** [comment]
You raise a valid point regarding closures. Indeed, my first implementation of all these functions was as closures, before I refactored the code to use the object-oriented design.

The advantages of the class-based design are:
- It doesn't make a new copy of the byte code every time you make a blocking function
- It would make it more straightforward to port the code to C should we wish to optimise it
- It’s what Python does in the standard library (`functools.partial` is actually a class that implements `__call__`; same with `operator.itemgetter`).

That said, I admit I’m probably overthinking this. Still, I’d like to merge this PR now (into a separate branch, not master), so I can work on integrating the new API with the entity service.
